usr/lib/common/profile_obj.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2020
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "pkcs11types.h"
+#include "defs.h"
+#include "host_defs.h"
+#include "h_extern.h"
+#include "trace.h"
+
+CK_RV profile_object_check_required_attributes(TEMPLATE *tmpl, CK_ULONG mode)
+{
+    /* CKO_PROFILE has no required attributes */
+    return template_check_required_base_attributes(tmpl, mode);
+}
+
+CK_RV profile_object_set_default_attributes(TEMPLATE *tmpl, CK_ULONG mode)
+{
+    CK_ATTRIBUTE *class_attr = NULL;
+    CK_ATTRIBUTE *profile_id_attr = NULL;
+
+    UNUSED(mode);
+
+    class_attr =
+        (CK_ATTRIBUTE *)malloc(sizeof(CK_ATTRIBUTE) + sizeof(CK_OBJECT_CLASS));
+    profile_id_attr =
+        (CK_ATTRIBUTE *)malloc(sizeof(CK_ATTRIBUTE) + sizeof(CK_PROFILE_ID));
+
+    if (!class_attr || !profile_id_attr) {
+        if (class_attr)
+            free(class_attr);
+        if (profile_id_attr)
+            free(profile_id_attr);
+
+        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+        return CKR_HOST_MEMORY;
+    }
+
+    class_attr->type = CKA_CLASS;
+    class_attr->ulValueLen = sizeof(CK_OBJECT_CLASS);
+    class_attr->pValue = (CK_BYTE *)class_attr + sizeof(CK_ATTRIBUTE);
+    *(CK_OBJECT_CLASS *)class_attr->pValue = CKO_PROFILE;
+
+    profile_id_attr->type = CKA_PROFILE_ID;
+    profile_id_attr->ulValueLen = sizeof(CK_PROFILE_ID);
+    profile_id_attr->pValue = (CK_BYTE *)profile_id_attr + sizeof(CK_ATTRIBUTE);;
+    *(CK_PROFILE_ID *)profile_id_attr->pValue = CKP_INVALID_ID; //XXX
```

Comment:
What does the XXX mean? Is this a left over TODO ?

Reply:
I was not sure what a reasonable default value is here, since the pkcs11 spec doesnt tell me, so this is kind of a question mark to reviewers.

Reply:
.. such that an application could create such an object and see what profile the lib conforms to based on the contents of the objct ? Thats another way of looking at it..

Reply:
Yes, but the base spec says "A given token can contain more than one profile ID" and which makes sense since a token could conform to multiple profiles. So i dont know how to make sense of it ..

Reply:
My interpretation was that a token should contain a profile object for every profile it conforms to but this is no where stated explicitly.

Reply:
right.... Not really well defined by PKCS#11 3.0 .....
So if profile objects are to be only created by the token itself, then you should actually hinder an application from creating one, I guess. So profile objects would be kind of read-only objects.

Reply:
hmm the spec doesnt say those objects cannot be created , copyed etc by an application either.

Reply:
So i implemented support for profile objects analogous to other object classes for now since the spec does not seem to treat them special. If it turns out tokens are required to come with r/o profile objects to conform to the baseline provider profile, i think its better to fix this as a bug than to publish something now that is not in the specs.

I added subtests for profile objects to the obj_mgmt_tests testcase.

Reply:
Agreed. Lets keep it as is for now.

If a token is supposed to contain these profile objects, and they are supposed to be found using C_FindObject, then the token should not only hinder an application from making changes to it, but it should also hinder an application from destroying/deleting such objects.  Otherwise one application could delete them and a second application would not find them anymore....
But that's a lot of guessing, so I would expect the PKCS #11 3.0 standard to specify such things more explicitly.